[Platformer] The player can move on platforms moving up

D8H opened the pull request:
Commited over:
* https://github.com/4ian/GDevelop/pull/2602
* https://github.com/4ian/GDevelop/pull/2603

This is a custom build to let Silver-Streak test it before the review:
* 1st version: https://www.dropbox.com/s/1l52e0vfm3ki1o0/gdevelop-plateformer-fix-2021-04-23.7z?dl=1
* 2nd version: https://www.dropbox.com/s/gxymf3zfs6njo7n/gdevelop-plateformer-fix-2021-04-25.7z?dl=1

It tries to fix:
* https://github.com/4ian/GDevelop/issues/897

# In master
* jumpthru platform
   * up: the player can't move and there is a gap smaller than 1 pixel
   * down: there is a gap smaller than 1 pixel
* regular platform
   * up: the player is jumpy
   * down: there is a gap smaller than 1 pixel

## Jumpthru platform going up
The player can't move right and left when the platform moves up because the player is in collision with the platform when moveX is called.
* doStepPreEvents() separate won't move the object up because it's a jumpthru
* beforeMovingX() set requestedDeltaY
* moveX() does move the player
* onFloor.beforeMovingY() reverts what moveX() did because the "slop is too step" 
* onFloor.beforeMovingY() applies the requestedDeltaY that make the player less than one pixel on top of the platform

## Regular platform going up
The player goes too much high and fell.
* doStepPreEvents() separate move the object up
* beforeMovingX() set requestedDeltaY according to the platform movement
* moveX() moves the player
* onFloor.beforeMovingY() applies the requestedDeltaY even though it's not needed
* _checkTransitionOnFloorOrFalling() more than 1 pixel apart, transition to falling 

# The fix
To fix these 2 "going up" issues, the plateform deltaY can be applied at the beginning.
The small gap is just that onFloor.beforeMovingY() make the player nearer to the platform by 1 pixel steps. A separate solve this but separate has rounding errors so the object can stay inside the platform (like 1e-15) so I added a collision check with an error margin in RuntimObject.

Silver-Streak commented:
@D8H 

So, seems a complete fix for my issue, as this solves: https://github.com/4ian/GDevelop/issues/2562
For my issue (#2562), I am no longer able to reproduce the jitter behavior at the start of the scene, I'm pretty sure you've solved this one.

But it does not fully solve: https://github.com/4ian/GDevelop/issues/897

Here's a video: https://streamable.com/i3onus

You can see that I can move on the platform, but the character still jitters/bounces up and down pretty rapidly as the platform raises or lowers. The example project is linked in the issue's first post. That said, I think you're connected with this already?

I remember you mentioning that there is a direct correlation to max fall speed, so I tested by setting the fall speed to 10000, which is above the acceleration (1500) in the example. This **does** eliminate the jitter when moving up/down, and at bare minimum establishes a workaround for the issue (we can instruct users to set up elevators/raising platform events to include an action that increases their max fall speed to a very large number while in collision, then reduces it back to normal when they've left the platform). 

Your changes do completely fix the inability to move on the raising/falling platform, though. No matter the fall speed I can still move now.

D8H commented:
> You can see that I can move on the platform, but the character still jitters/bounces up and down pretty rapidly as the platform raises or lowers. The example project is linked in the issue's first post. That said, I think you're connected with this already?

From what I see from the tests the player follow the platform, but maybe I don't have the right settings to reproduce. Are you using pixel rounding? In this case, try to round the platform position after moving it to see if it fix the issue.

Without pixel rounding:
```
LOG: 'playerY: -5.899999999999999 platformY: 14.1'
LOG: 'playerY: 18.200000000000003 platformY: 38.2'
LOG: 'playerY: 42.300000000000004 platformY: 62.300000000000004'
LOG: 'playerY: 66.4 platformY: 86.4'
LOG: 'playerY: 90.5 platformY: 110.5'
```

With pixel rounding:
```
LOG: 'playerY: -6 platformY: 14.1'
LOG: 'playerY: 18 platformY: 38.2'
LOG: 'playerY: 42 platformY: 62.300000000000004'
LOG: 'playerY: 66 platformY: 86.4'
LOG: 'playerY: 90 platformY: 110.5'
```

Silver-Streak commented:
I wonder if a fix could be an eventual "Elevator/Moving Platform" type for the main platform behavior that takes this into accounts, similar to jump through/normal/etc.

Regardless, your fix *does* give us an option for workarounds. Temporarily increasing max fall speed beyond acceleration is an easy thing to set up, and one I can document for users on the Wiki, if there isn't a more easily managed fix.

D8H commented:
> I wonder if a fix could be an eventual "Elevator/Moving Platform" type for the main platform behavior that takes this into accounts, similar to jump through/normal/etc.

The platform following could be move to doStepPostEvents to avoid the gap at rending. I will try it when the bug is fixed.

> Regardless, your fix _does_ give us an option for workarounds. Temporarily increasing max fall speed beyond acceleration is an easy thing to set up, and one I can document for users on the Wiki, if there isn't a more easily managed fix.

I don't know the effect of these properties very well, but in case of the platform following the limit I set is only between platform speed and max fall speed, acceleration shouldn't mater. Only really fast platforms will go beyond max fall speed.

D8H commented:
Sometimes the player can go through the jumpthru from above. It's a bug that already exist in the stable version. I think I'll leave it to another PR.

D8H commented:
I think I fixed the gap. I made a new build: https://www.dropbox.com/s/gxymf3zfs6njo7n/gdevelop-plateformer-fix-2021-04-25.7z?dl=1

Silver-Streak commented:
> I think I fixed the gap. I made a new build: https://www.dropbox.com/s/gxymf3zfs6njo7n/gdevelop-plateformer-fix-2021-04-25.7z?dl=1

While it is still not quite perfect without modifying max fall speed > acceleration, it is ALMOST unnoticeable now. 
Video of the new build, without modifying max fall speed: https://streamable.com/xs9avs

Video of the new build, with modifying max fall speed to 10000: https://streamable.com/0019pf

D8H commented:
I don't see the difference between the 2 captures. A red sprite and 4:2:0 is not good to notice a gap smaller than one pixel. Black and white should be better to compare as luminance is in full resolution in videos.

https://user-images.githubusercontent.com/2611977/115993351-aa12b500-a5d2-11eb-84d4-1d26dec3258e.mp4


I capture this with Max falling speed = 1000, Acceleration = 1500 and platform speed = 100

D8H commented:
When a jumpthru is moving up the player may not be able to land on it because it overlaps him a little. This is not a regression.

4ian commented:
I've not followed really what was done here, but originally the logic for following platforms on the X axis was to update `_requestedDeltaX` to be the same as the floor movement, and same for Y.
Y never really worked properly. Would you be able to tell exactly the strategy to fix this?

D8H commented:
# In master
* jumpthru platform
   * up: the player can't move and there is a gap smaller than 1 pixel
   * down: there is a gap smaller than 1 pixel
* regular platform
   * up: the player is jumpy
   * down: there is a gap smaller than 1 pixel

## Jumpthru platform going up
The player can't move right and left when the platform moves up because the player is in collision with the platform when moveX is called.
* doStepPreEvents() separate won't move the object up because it's a jumpthru
* beforeMovingX() set requestedDeltaY
* moveX() does move the player
* onFloor.beforeMovingY() reverts what moveX() did because the "slop is too step" 
* onFloor.beforeMovingY() applies the requestedDeltaY that make the player less than one pixel on top of the platform

## Regular platform going up
The player goes too much high and fell.
* doStepPreEvents() separate move the object up
* beforeMovingX() set requestedDeltaY according to the platform movement
* moveX() moves the player
* onFloor.beforeMovingY() applies the requestedDeltaY even though it's not needed
* _checkTransitionOnFloorOrFalling() more than 1 pixel apart, transition to falling 

# The fix
To fix these 2 "going up" issues, the plateform deltaY can be applied at the beginning.
The small gap is just that onFloor.beforeMovingY() make the player nearer to the platform by 1 pixel steps. A separate solve this but separate has rounding errors so the object can stay inside the platform (like 1e-15) so I added a collision check with an error margin in RuntimObject.

Silver-Streak commented:
@D8H @4ian Is there anything particular we're waiting on with this one?

I'm seeing all of the separate PRs, do they replace the need for this PR, and then just need the review? Or are they separate?

D8H commented:
> 
> 
> @D8H @4ian Is there anything particular we're waiting on with this one?
> 
> I'm seeing all of the separate PRs, do they replace the need for this PR, and then just need the review? Or are they separate?

I split the PR in 3, this one is now only about the platforms issues but it needs the 2 other PR to work. I updated the description.
About the next step, there is a design choice to to make. I made a suggestion here: https://github.com/4ian/GDevelop/pull/2602#discussion_r625042873 but I'm not expecting a quick answer (there were big merges recently).

4ian commented:
Sorry I felt behind in these PR. I'll try to take a look today :)

D8H commented:
This original issue is solved by:
* https://github.com/4ian/GDevelop/pull/3009
